src/helics/core/GlobalTimeCoordinator.cpp

```diff
@@ -0,0 +1,273 @@
+{
+    auto ret = MessageProcessingResult::CONTINUE_PROCESSING;
+
+    if (!dependencies.checkIfReadyForExecEntry(false, false)) {
+        bool allowed{false};
+        if (currentTimeState == TimeState::exec_requested_iterative) {
+            allowed = true;
+            for (auto& dep : dependencies) {
+                if (dep.dependency) {
+                    if (dep.minFed != mSourceId) {
+                        allowed = false;
+                        break;
+                    }
+                    if (dep.responseSequenceCounter != sequenceCounter) {
+                        allowed = false;
+                        break;
+                    }
+                }
+            }
+        }
+        if (!allowed) {
+            return ret;
+        }
+    }
+    executionMode = true;
+    ret = MessageProcessingResult::NEXT_STEP;
+
+    currentMinTime = timeZero;
+    currentTimeState = TimeState::time_granted;
+    nextEvent = timeZero;
+
+    ActionMessage execgrant(CMD_EXEC_GRANT);
+    execgrant.source_id = mSourceId;
+    transmitTimingMessagesDownstream(execgrant);
+    transmitTimingMessagesUpstream(execgrant);
+    return ret;
+}
+
+void GlobalTimeCoordinator::transmitTimingMessagesUpstream(ActionMessage& msg) const
+{
+    if (!sendMessageFunction) {
+        return;
+    }
+
+    for (const auto& dep : dependencies) {
+        if (dep.connection == ConnectionType::child) {
+            continue;
+        }
+        if (!dep.dependent) {
+            continue;
+        }
+        msg.dest_id = dep.fedID;
+        if (msg.action() == CMD_EXEC_REQUEST) {
+            msg.setExtraDestData(dep.sequenceCounter);
+        }
+        sendMessageFunction(msg);
+    }
+}
+
+void GlobalTimeCoordinator::transmitTimingMessagesDownstream(ActionMessage& msg,
+                                                             GlobalFederateId skipFed) const
+{
+    if (!sendMessageFunction) {
+        return;
+    }
+    if ((msg.action() == CMD_TIME_REQUEST || msg.action() == CMD_TIME_GRANT)) {
+        for (const auto& dep : dependencies) {
+            if (dep.connection != ConnectionType::child) {
+                continue;
+            }
+            if (!dep.dependent) {
+                continue;
+            }
+            if (dep.fedID == skipFed) {
+                continue;
+            }
+            if (dep.dependency) {
+                if (dep.next > msg.actionTime) {
+                    continue;
+                }
+            }
+            msg.dest_id = dep.fedID;
+            sendMessageFunction(msg);
+        }
+    } else {
+        for (const auto& dep : dependencies) {
+            if (dep.dependent) {
+                if (dep.fedID == skipFed) {
+                    continue;
+                }
+                if (msg.action() == CMD_EXEC_REQUEST) {
+                    msg.setExtraDestData(dep.sequenceCounter);
+                }
+                msg.dest_id = dep.fedID;
+                sendMessageFunction(msg);
```

Comment:
Does a copy of `msg` get made, or will it be fully processed before a future iteration of the for loop alters the dest_id in the referenced message object?

Reply:
sendMessageFunction takes  const ref, so will copy it